### ISSUE TITLE
Process submitted to slurm doesn't fail properly with -with-trace

### ISSUE THREAD

**Issue by haqle314** [nextflow-io/nextflow#1425]
### Expected behavior and actual behavior

When a slurm job uses more memory than requested, it will be killed by the slurm daemon. When the `-with-trace` flag is given, the process exit code (content of `.exitcode`) is 0 instead of 1.

### Steps to reproduce the problem
```nextflow
// main.nf
nextflow.preview.dsl = 2

process consumeMemory {
    executor 'slurm'
    output: path "out"
    script:
    two_gb = 2 * (1024 * 3)
    """
    yes | tr \\\\n y | head -c ${two_gb} | grep y > out
    """
}

workflow {
    consumeMemory | view
}
```
Running this script without `-with-trace`, it fails as expected. With `-with-trace`, it succeeds.

### Environment 

* Nextflow version: 19.10.0 build 5170
* Java version: Groovy 2.5.8 on OpenJDK 64-Bit Server VM 1.8.0_152-release-1056-b12
* Operating system: Linux

### Additional context
Basically, the script's exit status got lost somewhere inside the shell function `nxf_trace_linux`.

**Comment by haqle314** [comment]
I tried some more variations and it seems that this bug is specific to when the process receives a signal from the slurm daemon. If the script itself returns a non-zero exit code, everything still works as expected.

**Comment by pditommaso** [comment]
The exit code should be captured by the `nxf_main_ret` variable 

https://github.com/nextflow-io/nextflow/blob/ee4d8aeed904261a33432fc45ec6a2395345fde6/modules/nextflow/src/test/groovy/nextflow/executor/test-bash-wrapper-with-trace.txt#L237

I don't think this is related to dsl2.

**Comment by haqle314** [comment]
Correct me if I'm wrong but I think I'm onto something. Slurm sees that too much memory is consumed, it sends SIGTERM to the process. In `.command.run`, when given `-with-trace`, the script wraps itself in an extra shell level where `$NXF_ENTRY` is `nxf_trace`. Since `trap on_term ...` is inside `nxf_main`, I suspect that this wrapper process doesn't handle the SIGTERM from slurm appropriately.

**Comment by pditommaso** [comment]
Yes, but main entry remain `nxf_main` which sets the trap. On `TERM`  the `nxf_kill` should kill all children. 

https://github.com/nextflow-io/nextflow/blob/ee4d8aeed904261a33432fc45ec6a2395345fde6/modules/nextflow/src/test/groovy/nextflow/executor/test-bash-wrapper-with-trace.txt#L246-L249

**Comment by haqle314** [comment]
Well then, I'm out of ideas. Is there anything I could provide to help you debugging this?

**Comment by tbugfinder** [comment]
Hi @haqle314 
I'd like to test this on my cluster, too, which is Slurm 19.05, though.
Where did you set the memory constraint?

**Comment by tbugfinder** [comment]
Hi @haqle314 
I've used `stress` to run this memory test and also set `memory = '1 GB'`.
IMHO, you have to be careful while using pipes within the script and not having something like `set -euxo pipefail` set.
In my case that setting ends up in consistent exit codes.

**Comment by haqle314** [comment]
Sorry, I should've specified, on the cluster that I'm using (slurm 18.08.7), it defaults to 1 GB memory if no specific value was requested. I just used that snippet because that's the first thing I found on stackoverflow about testing memory request :smile: . It looks like `stress` is not installed on my cluster though. I will try it again with `set -euxo pipefail`.

**Comment by haqle314** [comment]
I just noticed that in the example script, I used `*` instead of `**` :man_facepalming: so that was 6 KB instead of 2 GB. After fixing that, I found that the process always fail as you'd expect, something about broken pipe. But that's not really the same as the case that I'm describing, where a single process consumes too much memory and got killed, no piping was involved. I will figure out some script to simulate this and report back.
